src/transports/tcp/TcpConnectionAcceptor.cpp

```diff
@@ -84,25 +83,29 @@ folly::Future<folly::Unit> TcpConnectionAcceptor::start(
   serverSocket_.reset(
       new folly::AsyncServerSocket(serverThread_->getEventBase()));
 
-  return via(serverThread_->getEventBase()).then([this] {
-    folly::SocketAddress addr;
-    addr.setFromLocalPort(options_.port);
-
-    serverSocket_->bind(addr);
-
-    for (auto const& callback : callbacks_) {
-      serverSocket_->addAcceptCallback(callback.get(), callback->eventBase());
-    }
-
-    serverSocket_->listen(options_.backlog);
-    serverSocket_->startAccepting();
-
-    for (auto& i : serverSocket_->getAddresses()) {
-      LOG(INFO) << "Listening on " << i.describe();
-    }
-
-    return folly::unit;
-  });
+  // The AsyncServerSocket needs to be accessed from the listener thread only.
+  // This will propagate out any exceptions the listener throws.
+  folly::via(
+      serverThread_->getEventBase(),
+      [this] {
+        folly::SocketAddress addr;
+        addr.setFromLocalPort(options_.port);
+
+        serverSocket_->bind(addr);
+
+        for (auto const& callback : callbacks_) {
+          serverSocket_->addAcceptCallback(
+              callback.get(), callback->eventBase());
+        }
+
+        serverSocket_->listen(options_.backlog);
+        serverSocket_->startAccepting();
+
+        for (auto& i : serverSocket_->getAddresses()) {
+          LOG(INFO) << "Listening on " << i.describe();
+        }
+      })
+      .get();
```

Comment:
Can we use "wait" in here and do better logging?

Should we clear onAccept_ variable in case of failure, to be able to retry again?